File: original_recipe/bbcursive/src/main/java/bbcursive/lib/opt_.java
package bbcursive.lib;

import org.jetbrains.annotations.NotNull;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.function.UnaryOperator;

import static bbcursive.std.bb;

/**
 * Created by jim on 1/17/16.
 */
public enum opt_ {
    ;

    @NotNull
    public static UnaryOperator<ByteBuffer> opt(UnaryOperator<ByteBuffer>... unaryOperators) {
        return new ByteBufferUnaryOperator(unaryOperators);
    }

    public static class ByteBufferUnaryOperator implements UnaryOperator<ByteBuffer> {
        private final UnaryOperator<ByteBuffer>[] allOrPrevious;

        @NotNull
        @Override
        public String toString() {
            return "opt:" + Arrays.deepToString(allOrPrevious);
        }

        public ByteBufferUnaryOperator(UnaryOperator<ByteBuffer>[] allOrPrevious) {

            this.allOrPrevious = allOrPrevious;
        }

        @NotNull
        @Override
        public ByteBuffer apply(@NotNull ByteBuffer buffer) {
            int position = buffer.position();
            ByteBuffer r = bb(buffer, allOrPrevious);
            if (null == r) {
                buffer.position(position);
            }
            return buffer;
        }
    }
}

File: original_recipe/bbcursive/src/main/java/bbcursive/lib/opt_.kt
package bbcursive.lib

import bbcursive.func.UnaryOperator
import bbcursive.std
import java.nio.ByteBuffer
import java.util.*

/**
 * Created by jim on 1/17/16.
 */
internal object opt_ {
    @JvmStatic
    fun opt(
        vararg unaryOperators:UnaryOperator<ByteBuffer>
    ): UnaryOperator<ByteBuffer> {
        return ByteBufferUnaryOperator(*unaryOperators)
    }

    class ByteBufferUnaryOperator(vararg val allOrPrevious: UnaryOperator<ByteBuffer>) :
        UnaryOperator<ByteBuffer> {
        override fun toString(): String = "opt:" + Arrays.deepToString(allOrPrevious)
        override operator fun invoke(buffer: ByteBuffer): ByteBuffer {
            val position = buffer!!.position()
            val r = std.bb(buffer, *allOrPrevious  )
            if (null == r) buffer.position(position)
            return buffer
        }
    }
}